lib/galaxy/webapps/galaxy/api/workflows.py

```diff
@@ -812,13 +812,16 @@ def invoke(self, trans: GalaxyWebTransaction, workflow_id, payload, **kwd):
 
         require_exact_tool_versions = util.string_as_bool(payload.get("require_exact_tool_versions", "true"))
         tools = self.workflow_contents_manager.get_all_tools(workflow)
-        missing_tools = [
-            tool
-            for tool in tools
-            if not self.app.toolbox.has_tool(
-                tool["tool_id"], tool_version=tool["tool_version"], exact=require_exact_tool_versions
+        missing_tools = []
+        for tool_info in tools:
+            tool = self.app.toolbox.get_tool(
+                tool_info["tool_id"], tool_version=tool_info["tool_version"], exact=require_exact_tool_versions
             )
-        ]
+            if tool is None:
+                missing_tools.append(tool_info)
+            elif require_exact_tool_versions and tool_info["tool_version"] != tool.version:
+                # check version returned is correct because get_tool does not check tool_version even with exact=True
+                missing_tools.append(tool_info)
```

Comment:
This might come back to bite us after a few Galaxy releases with new tool version. It makes sense for TS tools, not so much for local tools I think.

Reply:
Hm, not sure I understand, do local tools not specify a tool version? Should I change something?

Reply:
ok, done